source/pip/src/interpreter/data_interop.rs

```diff
@@ -0,0 +1,432 @@
+            }
+            Ok(Value::Array(array.into()))
+        }
+        Ty::Udt(_, res) => {
+            let qsc::hir::Res::Item(item_id) = res else {
+                panic!("Udt should be an item");
+            };
+            let Some(udt) = ctx.udt_ty(item_id) else {
+                unreachable!(
+                    "we verified that the udt is defined in `first_unsupported_interop_ty`"
+                );
+            };
+
+            // Handle `Complex` special case.
+            if let Ok(v) = obj.extract::<num_complex::Complex64>(py) {
+                if is_complex_udt(udt) {
+                    let tuple = Value::Tuple(Rc::new([Value::Double(v.re), Value::Double(v.im)]));
+                    return Ok(tuple);
+                }
+            }
+
+            let udt_fields = obj.extract::<UdtFields>(py)?;
+
+            let mut tuple = Vec::new();
+            for (name, ty) in collect_udt_fields(udt)? {
+                let Some(value) = udt_fields.get(&*name) else {
+                    return Err(PyTypeError::new_err(format!(
+                        "missing field {} in {}",
+                        name, udt.name,
+                    )));
+                };
+                tuple.push(convert_value_ir_with_ty(ctx, py, value, ty)?);
+            }
+            Ok(Value::Tuple(tuple.into()))
+        }
+        _ => unimplemented!("input type: {ty}"),
+    }
+}
+
+pub(super) fn type_ir_from_qsharp_ty(ctx: &interpret::Interpreter, ty: &Ty) -> PyResult<TypeIR> {
+    match ty {
+        Ty::Prim(prim) => {
+            let prim = match prim {
+                Prim::Bool => PrimitiveKind::Bool,
+                Prim::Int | Prim::BigInt => PrimitiveKind::Int,
+                Prim::Double => PrimitiveKind::Double,
+                Prim::String => PrimitiveKind::String,
+                Prim::Pauli => PrimitiveKind::Pauli,
+                Prim::Result => PrimitiveKind::Result,
+
+                Prim::Qubit | Prim::Range | Prim::RangeTo | Prim::RangeFrom | Prim::RangeFull => {
+                    return Err(PyTypeError::new_err(format!(
+                        "unsupported interop type: `{ty}`"
+                    )));
+                }
+            };
+            Ok(TypeIR::Primitive(prim))
+        }
+        Ty::Array(ty) => Ok(TypeIR::Array(vec![type_ir_from_qsharp_ty(ctx, ty)?])),
+        Ty::Tuple(items) => {
+            let mut tuple = Vec::new();
+            for item in items {
+                tuple.push(type_ir_from_qsharp_ty(ctx, item)?);
+            }
+            Ok(TypeIR::Tuple(tuple))
+        }
+        Ty::Udt(name, res) => {
+            let qsc::hir::Res::Item(item_id) = res else {
+                panic!("Udt should be an item");
+            };
+            let Some(udt) = ctx.udt_ty(item_id) else {
+                unreachable!(
+                    "we verified that the udt is defined in `first_unsupported_interop_ty`"
+                );
+            };
```

Comment:
Here the panic should be fine... these types came from inside the evaluator, so they should always line up.

Reply:
(feel free to resolve)